Client/Frontend/Browser/OpenPdfHelper.swift

```diff
@@ -0,0 +1,150 @@
+    static func canOpen(url: NSURL) -> Bool
+    func showOpenInView(inView parentView: UIView, forWebView webView: WKWebView)
+    func hideOpenInView(forWebView webView: WKWebView)
+    func open()
+}
+
+struct OpenInHelperFactory {
+    static func helperForURL(url: NSURL) -> OpenInHelper? {
+        if OpenPdfInHelper.canOpen(url) {
+            return OpenPdfInHelper(url: url)
+        }
+
+        return nil
+    }
+}
+
+class OpenPdfInHelper: NSObject, OpenInHelper, UIDocumentInteractionControllerDelegate {
+    private var view: OpenInView?
+    private var docController: UIDocumentInteractionController? = nil
+
+    private var parentView: UIView!
+    private var webView: WKWebView!
+
+    init(url: NSURL) {
+        super.init()
+        let contentsOfFile = NSData(contentsOfURL: url)
+        let dirPaths = NSSearchPathForDirectoriesInDomains(.DocumentDirectory,
+            .UserDomainMask, true)
+
+        let dirPath = NSURL(fileURLWithPath: dirPaths[0]).URLByAppendingPathComponent("pdfs")
+        let fileManager = NSFileManager.defaultManager()
+        do {
+            guard let lastPathComponent = url.lastPathComponent else {
+                log.error("failed to create proper URL")
+                return
+            }
+            let filePath = dirPath.URLByAppendingPathComponent(lastPathComponent)
+            try fileManager.createDirectoryAtURL(dirPath, withIntermediateDirectories: true, attributes: nil)
+            if fileManager.createFileAtPath(filePath.absoluteString, contents: contentsOfFile, attributes: nil) {
+                let openInURL = NSURL(fileURLWithPath: filePath.absoluteString)
+                docController = UIDocumentInteractionController(URL: openInURL)
+                docController?.delegate = self
+            } else {
+                log.error("Unable to create local version of PDF file at \(filePath)")
+            }
+        } catch {
+            log.error("Error on creating directory at \(dirPath)")
+        }
+    }
+    
+    static func canOpen(url: NSURL) -> Bool {
+        guard let pathExtension = url.pathExtension else { return false }
+        return pathExtension == FileType.PDF.rawValue && UIApplication.sharedApplication().canOpenURL(NSURL(string: "itms-books:")!)
+    }
+
+    func showOpenInView(inView parentView: UIView, forWebView webView: WKWebView) {
+        let overlayView = OpenInView()
+
+        overlayView.openInButton.addTarget(self, action: "open", forControlEvents: .TouchUpInside)
+        parentView.addSubview(overlayView)
+        overlayView.snp_makeConstraints { make in
+            make.trailing.equalTo(parentView)
+            make.leading.equalTo(parentView)
+            make.height.equalTo(OpenInViewUX.ViewHeight)
+            make.bottom.equalTo(webView.snp_top)
+        }
+
+        webView.snp_updateConstraints { make in
+            make.edges.equalTo(EdgeInsetsMake(OpenInViewUX.ViewHeight, left: 0, bottom: 0, right: 0))
+        }
```

Comment:
I would still strongly prefer to not have this component touching views and layouts outside of what it owns. Ideally, we would avoid touching `webView` constraints at all, and instead have the webView layout be defined relative to some header view that we could push views onto. With that approach, `OpenInHelper` could simply return a view that BVC pushes into this header view, and this might allow things to be handled more gracefully in the full-browser context (e.g., switching tabs). That might help deal with some of the aforementioned UI glitches.

I won't block on doing this if you don't want to change it, but curious to hear your thoughts.